### PR TITLE
Feature: State

### PR DESCRIPTION
**Pull request by bohdan-meltiukhov**
# State
This PR adds state and state-aware components. It also includes:
- State Manager,
- Actions,
- Mutators.

### CODE REVIEW

File: ui-components/tests/state/actions/get-files-action/index.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        mutator.apply(state);
+        assert.strictEqual(state.fileList, files, 'The GetFilesAction should provide the correct files.');
+      }
+    },
+  };
+
+  const action = new GetFilesAction();
+  action.apply(stateManager, apiService);
+
+  assert.timeout(1000);
```
Comment (dkalinin) [inline ui-components/tests/state/actions/get-files-action/index.js:42]: This is not correct. Let's discuss it on the meeting.

File: ui-components/app/components/file-list-page/index.js [hunk 2]
```diff
@@ -79,11 +82,23 @@ export default class FileListPage extends Component {
...
-    const apiService = ApiService.getInstance();
-    apiService.getFiles()
-      .then((files) => {
-        this.fileList.files = files;
-      });
+  /** @inheritdoc */
+  initState() {
+    this.onStateChanged('fileList', (event) => {
+      const state = event.detail.state;
+      this.fileList.files = state.fileList;
```
Comment (dkalinin) [inline ui-components/app/components/file-list-page/index.js:91]: What if fileList doesn't exist yet?

File: ui-components/app/components/file-list-page/index.js [hunk 3]
```diff
@@ -79,11 +82,23 @@ export default class FileListPage extends Component {
...
+    });
+
+    this.onStateChanged('isFileListLoading', (event) => {
+      const state = event.detail.state;
+      if (state.isFileListLoading) {
+        this.fileListContainer.innerHTML = '<div class="loader"></div>';
+      } else {
+        this.fileListContainer.innerHTML = '';
+        this.fileList = new FileList(this.fileListContainer);
```
Comment (dkalinin) [inline ui-components/app/components/file-list-page/index.js:100]: It is not a responsibility of `isFileListLoading` handler to create new FileList. It's better to do it in `fileList` handler.

File: ui-components/tests/state/actions/get-files-action/index.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import FileListMutator from '../../../../app/state/mutators/file-list-mutator';
+import GetFilesAction from '../../../../app/state/actions/get-files-action';
+
+const {module, test} = QUnit;
+
+module('The GetFilesAction');
+
+test('should mutate the stateManager.', (assert) => {
```
Comment (dkalinin) [inline ui-components/tests/state/actions/get-files-action/index.js:8]: should call mutate of state manager.

File: ui-components/tests/state/actions/get-files-action/index.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    },
+  ];
+
+  const apiServiceMock = {
+    getFiles: () => files,
+  };
+
+  const stateManagerMock = {
+    mutate: (mutator) => {
+      if (mutator instanceof FileListMutator) {
```
Comment (dkalinin) [inline ui-components/tests/state/actions/get-files-action/index.js:29]: This also should be checked in assert.

File: ui-components/tests/state/state-manager/index.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+const {module, test} = QUnit;
+
+module('The StateManager');
+
+test('should handle the state change.', (assert) => {
+  const state = {};
+  const stateManager = new StateManager(state, ApiService.getInstance());
+
+  const field = 'field';
```
Comment (dkalinin) [inline ui-components/tests/state/state-manager/index.js:12]: There's no need to have this variable.

- [ ] Removed.

File: ui-components/app/components/file-list-page/index.js (hunk 3 above)
Comment (bohdan-meltiukhov) [reply ui-components/app/components/file-list-page/index.js:100]: I changed it now.

File: ui-components/app/components/file-list-page/index.js (hunk 2 above)
Comment (bohdan-meltiukhov) [reply ui-components/app/components/file-list-page/index.js:91]: The fileList is initialized once in the initNestedComponents method now.

File: ui-components/tests/state/actions/get-files-action/index.js (hunk 4 above)
Comment (bohdan-meltiukhov) [reply ui-components/tests/state/actions/get-files-action/index.js]: I edited the test message.

File: ui-components/tests/state/actions/get-files-action/index.js (hunk 5 above)
Comment (bohdan-meltiukhov) [reply ui-components/tests/state/actions/get-files-action/index.js]: The test checks it now.

File: ui-components/tests/state/actions/get-files-action/index.js [hunk 7]
```diff
@@ -20,13 +23,23 @@ test('should mutate the stateManager.', (assert) => {
...
   const stateManagerMock = {
     mutate: (mutator) => {
-      if (mutator instanceof FileListMutator) {
+      if (mutator instanceof IsFileListLoadingMutator) {
+        isFileListLoadingMutatorProvided = true;
+      } else if (mutator instanceof FileListMutator) {
+        fileListMutatorProvided = true;
         assert.strictEqual(mutator._fileList, files, 'The GetFilesAction should create an instance of the ' +
           'FileListMutator with correct files.');
       }
```
Comment (dkalinin) [inline ui-components/tests/state/actions/get-files-action/index.js]: Let's refactor this to use assert.step().

File: ui-components/tests/state/actions/get-files-action/index.js [hunk 8]
```diff
@@ -31,15 +31,12 @@ test('should call the mutate method of the state manager.', async (assert) => {
...
     mutate: (mutator) => {
       if (mutator instanceof IsFileListLoadingMutator) {
-        isFileListLoadingMutatorProvided = true;
+        assert.step('isFileListLoadingMutator: ' + mutator._isLoading);
       } else if (mutator instanceof FileListMutator) {
-        fileListMutatorProvided = true;
+        assert.step('fileListMutator');
         assert.strictEqual(mutator._fileList, files, 'The GetFilesAction should create an instance of the ' +
           'FileListMutator with correct files.');
       }
```
Comment (dkalinin) [inline ui-components/tests/state/actions/get-files-action/index.js]: - [ ] Let's also verify that only expected mutators are called. Use `mutator.constructor.name`.

**Comment by dkalinin** [comment]
Please do not add any commit after PR is approved. After approval PR should be merged to target branch. If you want to add improvements then create a new branch and separate PR.

**Comment by bohdan-meltiukhov** [comment]
Got it. I will not add commits in such cases anymore.

**Pull request merged**